File: server/src/Models/DMToolkitModels/Instances/Character.cs
using DMToolkit.Models.Items.Bases;
using DMToolkit.Models.JoinTables;

namespace DMToolkit.Models.Instances;

public class Character : WorthBase
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public string Name { get; set; } = string.Empty;
    public string Alignment { get; set; } = string.Empty;
    public int Hp { get; set; } = 0;
    public int TempHp { get; set; } = 0;
    public List<CharacterCharacterClassInstance> CharacterCharacterClassInstances { get; set; } = new();
    public string BackgroundInstanceId { get; set; } = string.Empty;
    public BackgroundInstance? BackgroundInstance { get; set; } = null;
    public string SpeciesInstanceId { get; set; } = string.Empty;
    public SpeciesInstance? SpeciesInstance { get; set; } = null;
    public ICollection<AbilityScoreInstance> AbilityScoreInstances { get; set; } = new List<AbilityScoreInstance>();
    public string PhysicalDescription { get; set; } = string.Empty;
    public string Personality { get; set; } = string.Empty;
    public string Ideals { get; set; } = string.Empty;
    public string Bonds { get; set; } = string.Empty;
    public string Flaws { get; set; } = string.Empty;
    public int ProficiencyBonus { get; set; } = 0;
    public ICollection<CharacterItemBase> CharacterItemBases { get; set; } = new List<CharacterItemBase>();
    public ICollection<CharacterSpell> CharacterSpells { get; set; } = new List<CharacterSpell>();
}

File: web-client/src/lib/types/dm-tool-types/instances/character.ts
import type { ItemInstanceBase, ItemInstanceBaseDto } from "../items/bases/itemInstanceBase";
import type { Worth, WorthDto } from "../items/bases/worth";
import type { CharacterSpell, CharacterSpellDto } from "../relationships/characterSpell";
import type { AbilityScoreInstanceDto, AbilityScores } from "./abilityScoreInstance";
import type { BackgroundInstance, BackgroundInstanceDto } from "./backgroundInstance";
import type { CharacterClassInstance, CharacterClassInstanceDto } from "./characterClassInstance";
import type { SpeciesInstance, SpeciesInstanceDto } from "./speciesInstance";

export interface Character {
    id: string;
    name: string;
    alignment: string;
    hp: number;
    tempHp: number;
    primaryCharacterClassInstance: CharacterClassInstance | null;
    secondaryCharacterClassInstance: CharacterClassInstance | null;
    tertiaryCharacterClassInstance: CharacterClassInstance | null;
    backgroundInstance: BackgroundInstance | null;
    speciesInstance: SpeciesInstance | null;
    scores: AbilityScores;
    physicalDescription: string;
    personality: string;
    ideals: string;
    bonds: string;
    flaws: string;
    coins: Worth;
    inventory: ItemInstanceBase[];
    characterSpells: CharacterSpell[];
}

export interface CharacterDto {
    id: string;
    name: string;
    alignment: string;
    hp: number;
    tempHp: number;
    primaryCharacterClassInstanceDto: CharacterClassInstanceDto | null;
    secondaryCharacterClassInstanceDto: CharacterClassInstanceDto | null;
    tertiaryCharacterClassInstanceDto: CharacterClassInstanceDto | null;
    backgroundInstance: BackgroundInstanceDto | null;
    speciesInstance: SpeciesInstanceDto | null;
    scoreInstances: AbilityScoreInstanceDto[];
    physicalDescription: string;
    personality: string;
    ideals: string;
    bonds: string;
    flaws: string;
    coins: WorthDto;
    inventory: ItemInstanceBaseDto[];
    characterSpells: CharacterSpellDto[];
}